File: legacy-typescript/src/runtime/file_io.ts
import fs from "node:fs";
import { errorAt } from "../diagnostics/errors.js";
import { Type } from "../semantics/types.js";
import { DateValue, parseDateLiteral, dateToString, toDayNumber, fromDayNumber } from "../util/dates.js";
import { RuntimeValue, TextFileHandle, RandomFileHandle } from "./values.js";

export function openTextFile(path: string, mode: string, line: number): TextFileHandle {
  const upper = mode.toUpperCase();
  if (upper !== "READ" && upper !== "WRITE" && upper !== "APPEND") {
    throw errorAt("FileError", line, "Invalid file mode.");
  }
  if (upper === "READ") {
    const content = fs.existsSync(path) ? fs.readFileSync(path, "utf8") : "";
    const lines = content.length === 0 ? [] : content.split(/\r?\n/);
    return { kind: "TextFile", mode: "READ", path, lines, index: 0, buffer: [], open: true };
  }
  return { kind: "TextFile", mode: upper as "WRITE" | "APPEND", path, lines: [], index: 0, buffer: [], open: true };
}

export function closeTextFile(handle: TextFileHandle): void {
  if (!handle.open) return;
  handle.open = false;
  if (handle.mode === "WRITE") {
    fs.writeFileSync(handle.path, handle.buffer.join("\n"), "utf8");
  }
  if (handle.mode === "APPEND") {
    fs.appendFileSync(handle.path, handle.buffer.join("\n"), "utf8");
  }
}

export function readTextLine(handle: TextFileHandle, line: number): string {
  if (!handle.open || handle.mode !== "READ") {
    throw errorAt("FileError", line, "File not open for READ.");
  }
  if (handle.index >= handle.lines.length) {
    throw errorAt("FileError", line, "End of file.");
  }
  return handle.lines[handle.index++].trim();
}

export function writeTextLine(handle: TextFileHandle, text: string, line: number): void {
  if (!handle.open || (handle.mode !== "WRITE" && handle.mode !== "APPEND")) {
    throw errorAt("FileError", line, "File not open for WRITE/APPEND.");
  }
  handle.buffer.push(text);
}

export function eofText(handle: TextFileHandle): boolean {
  if (!handle.open || handle.mode !== "READ") return true;
  return handle.index >= handle.lines.length;
}

export function openRandomFile(path: string): RandomFileHandle {
  const buffer = fs.existsSync(path) ? fs.readFileSync(path) : Buffer.alloc(0);
  return { kind: "RandomFile", path, position: 1, buffer, open: true };
}

export function closeRandomFile(handle: RandomFileHandle): void {
  if (!handle.open) return;
  handle.open = false;
  fs.writeFileSync(handle.path, handle.buffer);
}

export function seekRandom(handle: RandomFileHandle, pos: number, line: number): void {
  if (!handle.open) throw errorAt("FileError", line, "File not open.");
  if (pos < 1) throw errorAt("RangeError", line, "SEEK address out of range.");
  handle.position = pos;
}

export function getRecord(handle: RandomFileHandle, type: Type, line: number): RuntimeValue {
  if (!handle.open) throw errorAt("FileError", line, "File not open.");
  const size = sizeOfType(type, line);
  const offset = (handle.position - 1) * size;
  if (offset + size > handle.buffer.length) {
    throw errorAt("FileError", line, "Record out of range.");
  }
  const slice = handle.buffer.subarray(offset, offset + size);
  return decodeValue(type, slice, line);
}

export function putRecord(handle: RandomFileHandle, type: Type, value: RuntimeValue, line: number): void {
  if (!handle.open) throw errorAt("FileError", line, "File not open.");
  const size = sizeOfType(type, line);
  const offset = (handle.position - 1) * size;
  const needed = offset + size;
  if (handle.buffer.length < needed) {
    const newBuf = Buffer.alloc(needed);
    handle.buffer.copy(newBuf);
    handle.buffer = newBuf;
  }
  const encoded = encodeValue(type, value, line);
  encoded.copy(handle.buffer, offset);
}

export function sizeOfType(type: Type, line: number): number {
  switch (type.kind) {
    case "Integer":
      return 4;
    case "Real":
      return 8;
    case "Boolean":
      return 1;
    case "Char":
      return 4;
    case "Date":
      return 4;
    case "Enum":
      return 4;
    case "Array": {
      const count = type.bounds.reduce((acc, b) => acc * (b.high - b.low + 1), 1);
      return count * sizeOfType(type.elementType, line);
    }
    case "Record": {
      let total = 0;
      for (const field of type.fields.values()) {
        total += sizeOfType(field, line);
      }
      return total;
    }
    default:
      throw errorAt("TypeError", line, "Type not supported in RANDOMFILE.");
  }
}

function encodeValue(type: Type, value: RuntimeValue, line: number): Buffer {
  const size = sizeOfType(type, line);
  const buf = Buffer.alloc(size);
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  encodeInto(view, 0, type, value, line);
  return buf;
}

function decodeValue(type: Type, buf: Buffer, line: number): RuntimeValue {
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const { value } = decodeFrom(view, 0, type, line);
  return value;
}

function encodeInto(view: DataView, offset: number, type: Type, value: RuntimeValue, line: number): number {
  switch (type.kind) {
    case "Integer":
      view.setInt32(offset, value.data as number, true);
      return offset + 4;
    case "Real":
      view.setFloat64(offset, value.data as number, true);
      return offset + 8;
    case "Boolean":
      view.setUint8(offset, (value.data as boolean) ? 1 : 0);
      return offset + 1;
    case "Char":
      view.setUint32(offset, (value.data as string).codePointAt(0) ?? 0, true);
      return offset + 4;
    case "Date": {
      const d = value.data as DateValue;
      view.setInt32(offset, toDayNumber(d), true);
      return offset + 4;
    }
    case "Enum":
      view.setInt32(offset, value.data as number, true);
      return offset + 4;
    case "Array": {
      const arr = value.data as RuntimeValue[] | RuntimeValue[][];
      const count = type.bounds.length;
      const dims = type.bounds.map((b) => b.high - b.low + 1);
      const write = (data: any, depth: number, off: number): number => {
        if (depth === count - 1) {
          for (let i = 0; i < dims[depth]; i += 1) {
            off = encodeInto(view, off, type.elementType, data[i], line);
          }
          return off;
        }
        for (let i = 0; i < dims[depth]; i += 1) {
          off = write(data[i], depth + 1, off);
        }
        return off;
      };
      return write(arr, 0, offset);
    }
    case "Record": {
      const fields = value.data as Map<string, RuntimeValue>;
      for (const [name, fieldType] of type.fields.entries()) {
        const fieldValue = fields.get(name) as RuntimeValue;
        offset = encodeInto(view, offset, fieldType, fieldValue, line);
      }
      return offset;
    }
    default:
      throw errorAt("TypeError", line, "Type not supported in RANDOMFILE.");
  }
}

function decodeFrom(view: DataView, offset: number, type: Type, line: number): { value: RuntimeValue; offset: number } {
  switch (type.kind) {
    case "Integer": {
      const value = view.getInt32(offset, true);
      return { value: { type, data: value }, offset: offset + 4 };
    }
    case "Real": {
      const value = view.getFloat64(offset, true);
      return { value: { type, data: value }, offset: offset + 8 };
    }
    case "Boolean": {
      const value = view.getUint8(offset) === 1;
      return { value: { type, data: value }, offset: offset + 1 };
    }
    case "Char": {
      const code = view.getUint32(offset, true);
      const value = String.fromCodePoint(code);
      return { value: { type, data: value }, offset: offset + 4 };
    }
    case "Date": {
      const num = view.getInt32(offset, true);
      const value = fromDayNumber(num);
      return { value: { type, data: value }, offset: offset + 4 };
    }
    case "Enum": {
      const ordinal = view.getInt32(offset, true);
      return { value: { type, data: ordinal }, offset: offset + 4 };
    }
    case "Array": {
      const dims = type.bounds.map((b) => b.high - b.low + 1);
      const read = (depth: number, off: number): { data: any; offset: number } => {
        if (depth === dims.length - 1) {
          const arr: RuntimeValue[] = [];
          for (let i = 0; i < dims[depth]; i += 1) {
            const res = decodeFrom(view, off, type.elementType, line);
            arr.push(res.value);
            off = res.offset;
          }
          return { data: arr, offset: off };
        }
        const arr: any[] = [];
        for (let i = 0; i < dims[depth]; i += 1) {
          const res = read(depth + 1, off);
          arr.push(res.data);
          off = res.offset;
        }
        return { data: arr, offset: off };
      };
      const res = read(0, offset);
      return { value: { type, data: res.data }, offset: res.offset };
    }
    case "Record": {
      const fields = new Map<string, RuntimeValue>();
      for (const [name, fieldType] of type.fields.entries()) {
        const res = decodeFrom(view, offset, fieldType, line);
        fields.set(name, res.value);
        offset = res.offset;
      }
      return { value: { type, data: fields }, offset };
    }
    default:
      throw errorAt("TypeError", line, "Type not supported in RANDOMFILE.");
  }
}

File: csharp/Prescribe.Core/Runtime/FileIo.cs
using System.Buffers.Binary;
using Prescribe.Core.Diagnostics;
using Prescribe.Core.Semantics;
using Prescribe.Core.Util;

namespace Prescribe.Core.Runtime;

public interface IFileSystem
{
    bool Exists(string path);
    string ReadAllText(string path);
    void WriteAllText(string path, string content);
    void AppendAllText(string path, string content);
    byte[] ReadAllBytes(string path);
    void WriteAllBytes(string path, byte[] bytes);
}

public sealed class FileIo
{
    private readonly IFileSystem _fs;

    public FileIo(IFileSystem fs)
    {
        _fs = fs;
    }

    public TextFileHandle OpenTextFile(string path, string mode, int line)
    {
        var upper = mode.ToUpperInvariant();
        if (upper is not ("READ" or "WRITE" or "APPEND"))
        {
            throw Errors.At(ErrorType.FileError, line, "Invalid file mode.");
        }
        if (upper == "READ")
        {
            var content = _fs.Exists(path) ? _fs.ReadAllText(path) : "";
            var lines = content.Length == 0 ? new List<string>() : content.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None).ToList();
            return new TextFileHandle { Mode = TextFileMode.Read, Path = path, Lines = lines, Index = 0, Buffer = new List<string>(), Open = true };
        }
        return new TextFileHandle { Mode = upper == "WRITE" ? TextFileMode.Write : TextFileMode.Append, Path = path, Lines = new List<string>(), Index = 0, Buffer = new List<string>(), Open = true };
    }

    public void CloseTextFile(TextFileHandle handle)
    {
        if (!handle.Open) return;
        handle.Open = false;
        var text = string.Join("\n", handle.Buffer);
        if (handle.Mode == TextFileMode.Write)
        {
            _fs.WriteAllText(handle.Path, text);
        }
        else if (handle.Mode == TextFileMode.Append)
        {
            _fs.AppendAllText(handle.Path, text);
        }
    }

    public string ReadTextLine(TextFileHandle handle, int line)
    {
        if (!handle.Open || handle.Mode != TextFileMode.Read)
        {
            throw Errors.At(ErrorType.FileError, line, "File not open for READ.");
        }
        if (handle.Index >= handle.Lines.Count)
        {
            throw Errors.At(ErrorType.FileError, line, "End of file.");
        }
        return handle.Lines[handle.Index++].Trim();
    }

    public void WriteTextLine(TextFileHandle handle, string text, int line)
    {
        if (!handle.Open || (handle.Mode != TextFileMode.Write && handle.Mode != TextFileMode.Append))
        {
            throw Errors.At(ErrorType.FileError, line, "File not open for WRITE/APPEND.");
        }
        handle.Buffer.Add(text);
    }

    public bool EofText(TextFileHandle handle)
    {
        if (!handle.Open || handle.Mode != TextFileMode.Read) return true;
        return handle.Index >= handle.Lines.Count;
    }

    public RandomFileHandle OpenRandomFile(string path)
    {
        var buffer = _fs.Exists(path) ? _fs.ReadAllBytes(path) : Array.Empty<byte>();
        return new RandomFileHandle { Path = path, Position = 1, Buffer = buffer, Open = true };
    }

    public void CloseRandomFile(RandomFileHandle handle)
    {
        if (!handle.Open) return;
        handle.Open = false;
        _fs.WriteAllBytes(handle.Path, handle.Buffer);
    }

    public void SeekRandom(RandomFileHandle handle, int pos, int line)
    {
        if (!handle.Open) throw Errors.At(ErrorType.FileError, line, "File not open.");
        if (pos < 1) throw Errors.At(ErrorType.RangeError, line, "SEEK address out of range.");
        handle.Position = pos;
    }

    public RuntimeValue GetRecord(RandomFileHandle handle, TypeSymbol type, int line)
    {
        if (!handle.Open) throw Errors.At(ErrorType.FileError, line, "File not open.");
        var size = SizeOfType(type, line);
        var offset = (handle.Position - 1) * size;
        if (offset + size > handle.Buffer.Length)
        {
            throw Errors.At(ErrorType.FileError, line, "Record out of range.");
        }
        var slice = handle.Buffer.AsSpan(offset, size).ToArray();
        return DecodeValue(type, slice, line);
    }

    public void PutRecord(RandomFileHandle handle, TypeSymbol type, RuntimeValue value, int line)
    {
        if (!handle.Open) throw Errors.At(ErrorType.FileError, line, "File not open.");
        var size = SizeOfType(type, line);
        var offset = (handle.Position - 1) * size;
        var needed = offset + size;
        if (handle.Buffer.Length < needed)
        {
            var newBuf = new byte[needed];
            Buffer.BlockCopy(handle.Buffer, 0, newBuf, 0, handle.Buffer.Length);
            handle.Buffer = newBuf;
        }
        var encoded = EncodeValue(type, value, line);
        Buffer.BlockCopy(encoded, 0, handle.Buffer, offset, size);
    }

    public static int SizeOfType(TypeSymbol type, int line)
    {
        switch (type.Kind)
        {
            case TypeKind.Integer:
                return 4;
            case TypeKind.Real:
                return 8;
            case TypeKind.Boolean:
                return 1;
            case TypeKind.Char:
                return 4;
            case TypeKind.Date:
                return 4;
            case TypeKind.Enum:
                return 4;
            case TypeKind.Array:
            {
                var arrayType = (ArrayType)type;
                var count = arrayType.Bounds.Aggregate(1, (acc, b) => acc * (b.High - b.Low + 1));
                return count * SizeOfType(arrayType.ElementType, line);
            }
            case TypeKind.Record:
            {
                var recordType = (RecordType)type;
                var total = 0;
                foreach (var field in recordType.Fields.Values)
                {
                    total += SizeOfType(field, line);
                }
                return total;
            }
            default:
                throw Errors.At(ErrorType.TypeError, line, "Type not supported in RANDOMFILE.");
        }
    }

    private static byte[] EncodeValue(TypeSymbol type, RuntimeValue value, int line)
    {
        var size = SizeOfType(type, line);
        var buf = new byte[size];
        EncodeInto(buf, 0, type, value, line);
        return buf;
    }

    private static RuntimeValue DecodeValue(TypeSymbol type, byte[] buf, int line)
    {
        var res = DecodeFrom(buf, 0, type, line);
        return res.value;
    }

    private static int EncodeInto(byte[] buffer, int offset, TypeSymbol type, RuntimeValue value, int line)
    {
        switch (type.Kind)
        {
            case TypeKind.Integer:
                BinaryPrimitives.WriteInt32LittleEndian(buffer.AsSpan(offset, 4), (int)value.Data!);
                return offset + 4;
            case TypeKind.Real:
            {
                var bits = BitConverter.DoubleToInt64Bits((double)value.Data!);
                BinaryPrimitives.WriteInt64LittleEndian(buffer.AsSpan(offset, 8), bits);
                return offset + 8;
            }
            case TypeKind.Boolean:
                buffer[offset] = (bool)value.Data! ? (byte)1 : (byte)0;
                return offset + 1;
            case TypeKind.Char:
            {
                var ch = (string)value.Data!;
                var code = ch.Length == 0 ? 0 : char.ConvertToUtf32(ch, 0);
                BinaryPrimitives.WriteUInt32LittleEndian(buffer.AsSpan(offset, 4), (uint)code);
                return offset + 4;
            }
            case TypeKind.Date:
            {
                var d = (DateValue)value.Data!;
                BinaryPrimitives.WriteInt32LittleEndian(buffer.AsSpan(offset, 4), DateUtil.ToDayNumber(d));
                return offset + 4;
            }
            case TypeKind.Enum:
                BinaryPrimitives.WriteInt32LittleEndian(buffer.AsSpan(offset, 4), (int)value.Data!);
                return offset + 4;
            case TypeKind.Array:
            {
                var arrayType = (ArrayType)type;
                var dims = arrayType.Bounds.Select(b => b.High - b.Low + 1).ToArray();
                object data = value.Data!;
                int Write(object current, int depth, int off)
                {
                    if (depth == dims.Length - 1)
                    {
                        var arr = (object[])current;
                        for (var i = 0; i < dims[depth]; i += 1)
                        {
                            off = EncodeInto(buffer, off, arrayType.ElementType, (RuntimeValue)arr[i], line);
                        }
                        return off;
                    }
                    var outer = (object[])current;
                    for (var i = 0; i < dims[depth]; i += 1)
                    {
                        off = Write(outer[i], depth + 1, off);
                    }
                    return off;
                }
                return Write(data, 0, offset);
            }
            case TypeKind.Record:
            {
                var recordType = (RecordType)type;
                var fields = (Dictionary<string, RuntimeValue>)value.Data!;
                foreach (var (name, fieldType) in recordType.Fields)
                {
                    var fieldValue = fields[name];
                    offset = EncodeInto(buffer, offset, fieldType, fieldValue, line);
                }
                return offset;
            }
            default:
                throw Errors.At(ErrorType.TypeError, line, "Type not supported in RANDOMFILE.");
        }
    }

    private static (RuntimeValue value, int offset) DecodeFrom(byte[] buffer, int offset, TypeSymbol type, int line)
    {
        switch (type.Kind)
        {
            case TypeKind.Integer:
            {
                var value = BinaryPrimitives.ReadInt32LittleEndian(buffer.AsSpan(offset, 4));
                return (RuntimeValues.Make(type, value), offset + 4);
            }
            case TypeKind.Real:
            {
                var bits = BinaryPrimitives.ReadInt64LittleEndian(buffer.AsSpan(offset, 8));
                var value = BitConverter.Int64BitsToDouble(bits);
                return (RuntimeValues.Make(type, value), offset + 8);
            }
            case TypeKind.Boolean:
            {
                var value = buffer[offset] == 1;
                return (RuntimeValues.Make(type, value), offset + 1);
            }
            case TypeKind.Char:
            {
                var code = BinaryPrimitives.ReadUInt32LittleEndian(buffer.AsSpan(offset, 4));
                var value = char.ConvertFromUtf32((int)code);
                return (RuntimeValues.Make(type, value), offset + 4);
            }
            case TypeKind.Date:
            {
                var num = BinaryPrimitives.ReadInt32LittleEndian(buffer.AsSpan(offset, 4));
                var value = DateUtil.FromDayNumber(num);
                return (RuntimeValues.Make(type, value), offset + 4);
            }
            case TypeKind.Enum:
            {
                var ordinal = BinaryPrimitives.ReadInt32LittleEndian(buffer.AsSpan(offset, 4));
                return (RuntimeValues.Make(type, ordinal), offset + 4);
            }
            case TypeKind.Array:
            {
                var arrayType = (ArrayType)type;
                var dims = arrayType.Bounds.Select(b => b.High - b.Low + 1).ToArray();
                (object data, int off) Read(int depth, int off)
                {
                    if (depth == dims.Length - 1)
                    {
                        var arr = new object[dims[depth]];
                        for (var i = 0; i < dims[depth]; i += 1)
                        {
                            var res = DecodeFrom(buffer, off, arrayType.ElementType, line);
                            arr[i] = res.value;
                            off = res.offset;
                        }
                        return (arr, off);
                    }
                    var outer = new object[dims[depth]];
                    for (var i = 0; i < dims[depth]; i += 1)
                    {
                        var res = Read(depth + 1, off);
                        outer[i] = res.data;
                        off = res.off;
                    }
                    return (outer, off);
                }
                var res2 = Read(0, offset);
                return (RuntimeValues.Make(type, res2.data), res2.off);
            }
            case TypeKind.Record:
            {
                var recordType = (RecordType)type;
                var fields = new Dictionary<string, RuntimeValue>(StringComparer.Ordinal);
                foreach (var (name, fieldType) in recordType.Fields)
                {
                    var res = DecodeFrom(buffer, offset, fieldType, line);
                    fields[name] = res.value;
                    offset = res.offset;
                }
                return (RuntimeValues.Make(type, fields), offset);
            }
            default:
                throw Errors.At(ErrorType.TypeError, line, "Type not supported in RANDOMFILE.");
        }
    }
}
